lib/src/connector/connector.cc

```diff
@@ -561,14 +574,45 @@ void Connector::errorMessageCallback(const ParsedChunks& parsed_chunks) {
                       "request %2%",
                       error_id, cause_id);
             // The Associate Session request caused the error
-            session_association_.in_progress = false;
-            session_association_.success = false;
+            session_association_.got_messaging_failure = true;
             session_association_.error = description;
             session_association_.cond_var.notify_one();
         }
     }
 }
 
+// ttl_expired message callback
+
+void Connector::TTLMessageCallback(const ParsedChunks& parsed_chunks) {
+    assert(parsed_chunks.has_data);
+    assert(parsed_chunks.data_type == PCPClient::ContentType::Json);
+
+    auto ttl_msg_id = parsed_chunks.envelope.get<std::string>("id");
+    auto expired_msg_id = parsed_chunks.data.get<std::string>("id");
+
+    LOG_WARNING("Received TTL Expired message %1% from %2% related to message %3%",
+                ttl_msg_id, parsed_chunks.envelope.get<std::string>("sender"),
+                expired_msg_id);
+
+    if (TTL_expired_callback_)
+        TTL_expired_callback_(parsed_chunks);
+
+    if (session_association_.in_progress.load()) {
+        Util::lock_guard<Util::mutex> the_lock { session_association_.mtx };
+
+        if (!expired_msg_id.empty()
+                && expired_msg_id == session_association_.request_id) {
```

Comment:
Just wondering: Which component would be sending the TTL expired message in this instance?